### PR TITLE
Closed doors logic

### PR DESCRIPTION
**Pull request by wossnameGitHub**
New features:
1. Added algorithm to account for closed doors
This means that from now on, the following things will never happen again:
A) Levels being inaccessable due to closed doors
B) The player running into a dead end and therefore a softlock due to closed doors
Keep in mind that softlocking due to [missing items] is still completely possible, this closed doors algorithm is a whole different subject.

2. Added multiple small things to graph creation:
A) closed doors are marked in red, and point in the direction that is still traversable even if the door is closed
B) one-way-transitions are marked by using a dashed line
C) temp disabled exits are marked in black

In order for those to be added:
- Massive update/overhaul to entrance_rando.py
- Quite large update/overhaul to graph_creation.py

### CODE REVIEW

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 1]
```diff
@@ -352,69 +501,104 @@ def get_random_redirection(original: Transition, all_redirections: Iterable[Tran
...
+                try:
+                    find_and_break_open_connection(link_list)
+                except Exception as e:
+                    if str(e) == "NO CONNECTION FOUND":
+                        print()
+                        print("ATTENTION ATTENTION: THE RANDOMIZER CRASHED!!!")
+                        print("Please notify a developer!")
+                        print()
+                        break
+                    raise
```
Comment (Avasam) [inline Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:592]: It's better to create your own exception class than to rely on exception text.
Even more so, include that text in the exception message and it should show in the logs just the same once it bubbles-up, meaning you don't even need a try-except here.

Even if you still want to print separately for some reason, it'd look like this:
```suggestion
                try:
                    find_and_break_open_connection(link_list)
                except NoConnectionFoundError as error:
                    print(error)
                    raise
```

**Review by Avasam — CHANGES_REQUESTED** [review]
Went through a full review of `graph_creation`.
Only a very superficial look at the new rando logic, probably fine (the idea is sound, haven't checked the implementation in details yet).
The amount of global variables and context-specific top-level public methods we're using make me think that it might soon be time to think about making the "randomizer" it's own class.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 2]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
     return output_text
 
 
-def create_edges(transitions_map: Mapping[tuple[int, int], tuple[int, int]]):
-    connections = [(original[0], redirect[1]) for original, redirect in transitions_map.items()]
-    connections_two_way: list[tuple[int, int]] = []
-    connections_one_way: list[tuple[int, int]] = []
+def edge_text(
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:108]: renamed it to `edge_component`

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 3]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
     for pairing in connections_two_way:
-        output_text += (
-            f'<edge source="{TRANSITION_INFOS_DICT[pairing[0]].area_id}" '
-            + f'target="{TRANSITION_INFOS_DICT[pairing[1]].area_id}" isDirect="false" '
-            + f'id="{counter}"></edge>\n'
-        )
+        if pairing[1] in temp_disabled_exits:
+            output_text += edge_text(
+                pairing[0][0], pairing[1][1],
+                counter, Direction.TWOWAY, "#000000", LineType.SOLID,
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:168]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 4]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
-            + f'id="{counter}"></edge>\n'
+        output_text += edge_text(
+            pairing[0][0], pairing[1][1],
+            counter, Direction.ONEWAY, None, LineType.DASHED,
+        )
+        counter += 1
+    for pairing in connections_closed_door:
+        output_text += edge_text(
+            pairing[1][1], pairing[0][0],
+            counter, Direction.ONEWAY, "#ff0000", LineType.SOLID,
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:185]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 5]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
+    direct: Direction,
+    color: str | None,
+    line_type: LineType,
+):
+    direct_str = "true"
+    if direct == Direction.TWOWAY:
+        direct_str = "false"
+    output = (
+        f'<edge source="{transition_infos_dict_rando[start].area_id}" '
+        + f'target="{transition_infos_dict_rando[end].area_id}" isDirect="{direct_str}" '
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:121]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 6]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
+        + f'id="{counter}"'
+    )
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:123]: I chose the other option you provided.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 7]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
+    start: int,
+    end: int,
+    counter: int,
+    direct: Direction,
+    color: str | None,
+    line_type: LineType,
+):
+    direct_str = "true"
+    if direct == Direction.TWOWAY:
+        direct_str = "false"
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:118]: I implemented your second option.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 8]
```diff
@@ -23,11 +23,15 @@ class Choice(IntEnum):
...
-    LevelCRC.PENGUIN_TEMPLE,
-)
+class BucketType(IntEnum):
+    ORIGIN = 1
+    REDIRECT = 2
+
+
+class Priority(IntEnum):
+    CLOSED = 1
+    OPEN = 2
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:33]: Applied.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 9]
```diff
@@ -126,6 +191,7 @@ def create_edges(transitions_map: Mapping[tuple[int, int], tuple[int, int]]):
 def create_graphml(
     transitions_map: MutableMapping[tuple[int, int], tuple[int, int]],
     temp_disabled_exits: Sequence[tuple[int, int]],
+    closed_door_exits: list[tuple[int, int]],
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:194]: This line is now changed.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 10]
```diff
@@ -2,12 +2,25 @@
...
 
+
+class Direction(IntEnum):
+    ONEWAY = 1
+    TWOWAY = 2
+
+
+class LineType(IntEnum):
+    SOLID = 1
+    DASHED = 2
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:21]: Implemented your first option.

File: Dolphin scripts/Entrance Randomizer/__main__.py [hunk 11]
```diff
@@ -50,7 +51,7 @@
 
 # Dump spoiler logs and graph
 dump_spoiler_logs(starting_area_name, transitions_map, seed_string)
-create_graphml(transitions_map, temp_disabled_exits, seed_string, starting_area)
+create_graphml(transitions_map, temp_disabled_exits, closed_door_exits, seed_string, starting_area)
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/__main__.py:54]: I don't understand this comment.
That green line is already the current state of the code.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 12]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
+            output += "&quot;lineDash&quot;:&quot;2&quot;"
+        output += '}}"'
+    output += "></edge>\n"
+    return output
+
+
+def create_edges(
+    transitions_map: Mapping[tuple[int, int], tuple[int, int]],
+    temp_disabled_exits: list[tuple[int, int]],
+    closed_door_exits: list[tuple[int, int]],
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:140]: Applied.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 1 above)
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:592]: created a NoConnectionFoundError class

File: Dolphin scripts/Entrance Randomizer/lib/constants.py [hunk 13]
```diff
@@ -10,7 +10,7 @@
 from dolphin import memory  # pyright: ignore[reportMissingModuleSource]
 from lib.transition_infos import transition_infos
 
-__version__ = "0.4.0"
+__version__ = "0.4.5"
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/constants.py]: If you say so, but how am I supposed to name that entry?
Take this current merge request for instance, what should be the file name for the entry?
afbeelding

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 14]
```diff
@@ -93,31 +105,84 @@ def create_vertices(
...
+    )
+    if line_type == LineType.DASHED or color is not None:
+        output += ' ownStyles="{&quot;0&quot;:{'
+        if color is not None:
+            output += f"&quot;strokeStyle&quot;:&quot;{color}&quot;"
+            if line_type == LineType.DASHED:
+                output += ","
+        if line_type == LineType.DASHED:
+            output += "&quot;lineDash&quot;:&quot;2&quot;"
+        output += '}}"'
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py]: added create_own_style function.
Also, `2` is how the website decides what type of lineDash to use. There are multiple to choose from. The one I'm using currently is desginated `2`, so that's why I'm using 2 here.

File: Dolphin scripts/Entrance Randomizer/__main__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/__main__.py:54]: You had two generic typing invariance issues here. One with the `transition_map` parameter. One with the `closed_door_exits` parameter. But both fixes go in the `create_graphml` definition.

File: Dolphin scripts/Entrance Randomizer/lib/constants.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/constants.py]: `68.feature.md`

[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/constants.py]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 15]
```diff
@@ -352,69 +505,102 @@ def get_random_redirection(original: Transition, all_redirections: Iterable[Tran
...
+            key=lambda a: (
+                a.area_id in closed_door_levels, __connections_left[a.area_id],
+            ), reverse=True,
+        )
 
-        index = 2
+        global __current_hub, loose_ends, total_con_left
+
+        __current_hub = closed_door_levels[0]  # this list is shuffled, so just pick the first one
+        loose_ends = [*closed_door_exits]
```
Comment (Avasam) [inline Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:545]: You can create a list from an iterable, tuples are iterable. This is a bit better, even on Python 3.12 where list comprehensions are inlined
```suggestion
        loose_ends = list(closed_door_exits)
```

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 16]
```diff
@@ -352,69 +505,102 @@ def get_random_redirection(original: Transition, all_redirections: Iterable[Tran
...
-                # option 3: break open a connection that's part of a level loop,
-                # then restart this iteration
-                break_open_connection(level_list, index, link_list)
+                try:
+                    find_and_break_open_connection(link_list)
+                except NoConnectionFoundError:
+                    print(" ")
+                    print("ATTENTION ATTENTION: THE RANDOMIZER CRASHED!!!")
+                    print("Please notify a developer!")
+                    print(" ")
```
Comment (Avasam) [inline Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:593]: This message is applicable to more than just when no connections are found. This also currently suppresses the error and let the script still run despite being in a broken state, with the message hidden in logs, rather than having an obvious crash (ie: no HUD).

Or do you see a reason to let the script run despite this? (is there any use for a graceful failure/fallback here ?)

I feel like such a try-except belongs around the main loop, as an extra nudge that *if* you see this message in the logs, please send the traceback to the devs / open an issue.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 17]
```diff
@@ -352,69 +505,102 @@ def get_random_redirection(original: Transition, all_redirections: Iterable[Tran
...
 
-        link_list_to_transitions(
-            link_list,
-            transitions_map,
-            _possible_origins_bucket,
-            _possible_redirections_bucket,
-        )
+        # Once the link_list is completed, it's time to fill the transitions_map:
+        for link in link_list:
+            transitions_map[link[0]] = link[1]
```
Comment (Avasam) [inline Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:601]: A MutableMapping can be updated with an iterable of two-pairs (like a list or 2-tuples) OR anything that has key-values (like a dict)
```suggestion
        transitions_map.update(link_list)
```

**Review by Avasam — CHANGES_REQUESTED** [review]

File: Dolphin scripts/Entrance Randomizer/newsfragments/68.feature.md [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+Keep in mind that softlocking due to [missing items] is still completely possible, this closed doors algorithm is a whole different subject.
+
+**Added multiple small things to graph creation:**
+- closed doors are marked in red, and point in the direction that is still traversable even if the door is closed
+- one-way-transitions are marked by using a dashed line
+- temp disabled exits are marked in black
+
+In order for those to be added:
+- Massive update/overhaul to entrance_rando.py
+- Quite large update/overhaul to graph_creation.py
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/newsfragments/68.feature.md:16]: If you say so.

File: Dolphin scripts/Entrance Randomizer/newsfragments/68.feature.md [hunk 19]
```diff
@@ -0,0 +1,18 @@
+New features:
+
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/newsfragments/68.feature.md:2]: Sure I'll remove this line as well.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 20]
```diff
@@ -26,11 +39,26 @@
...
+
+def create_own_style(params: dict[str, str | None]):
+    output_text = ' ownStyles="{&quot;0&quot;:{'
+    first_key = True
+    for key, value in params.items():
+        if value is not None:
+            if not first_key:
+                output_text += ", "
+            output_text += f"&quot;{key}&quot;:&quot;{value}&quot;"
+            first_key = False
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:58]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 21]
```diff
@@ -1,12 +1,25 @@
 from __future__ import annotations
 
 from collections.abc import Mapping, Sequence
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:3]: I would have done so if you had told me to.
Anyway now it's added.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 22]
```diff
@@ -92,31 +109,79 @@ def create_vertices(
...
+        f'<edge source="{transition_infos_dict_rando[start].area_id}" '
+        + f'target="{transition_infos_dict_rando[end].area_id}" '
+        + f'isDirect="{direct_str}" '
+        + f'id="{counter}"'
+    )
+    if line_type == LineType.DASHED or color is not None:
+        output += create_own_style({
+            "strokeStyle": color,
+            "lineDash": "2" if line_type == LineType.DASHED else None,
+        })
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:131]: No it would not give the desired result if this is called with all values being None.
Yes I could rewrite the function to return an empty string if all values are None, but aren't we just moving the problem at that point?
Unless you can tell me how to cleanly do that check at the start of the function...

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 23]
```diff
@@ -352,69 +505,102 @@ def get_random_redirection(original: Transition, all_redirections: Iterable[Tran
 
 def set_transitions_map():  # noqa: PLR0915 # TODO: Break up in smaller functions
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:506]: Well yeah I can break it up as much as you want, but weren't you the one who wanted merge requests to be as small as possible?
If you want me to overhaul this python file even more than I already have, and you want me to do so in this very same MR... then sure I'll do that.
I would just like to get your confirmation on this part.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py [hunk 24]
```diff
@@ -54,7 +82,7 @@ def create_vertices(
         # The same logic applies to the Spirit Fights:
         # these will never appear on the map, therefore we remove the (Harry) suffix.
         area_name = (
-            TRANSITION_INFOS_DICT
+            transition_infos_dict_rando
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:85]: To name one thing it allows to display `Burning Outpost` and `Jungle Outpost`, both of which aren't present in `TRANSITION_INFOS_DICT`, as that dict considers them both to be equal to `Twin Outposts`.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 25]
```diff
@@ -197,152 +222,280 @@ def highjack_transition_rando():
...
-    for trans in ALL_POSSIBLE_TRANSITIONS:
-        if trans in one_way_exits or trans in disabled_exits:
-            ALL_POSSIBLE_TRANSITIONS_RANDO = [  # pyright: ignore[reportConstantRedefinition]
-                x for x in ALL_POSSIBLE_TRANSITIONS_RANDO if x != trans
-            ]
 
+def choose_random_exit(
+    level: int,
+    bucket_type: int,
+    priority: int,
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:328]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 26]
```diff
@@ -197,152 +222,280 @@ def highjack_transition_rando():
...
+):
+    from_exit = choose_random_exit(level_from, BucketType.ORIGIN, Priority.CLOSED)
+    to_entrance = choose_random_exit(level_to, BucketType.REDIRECT, Priority.OPEN)
 
-def unlink_two_levels(first: Area, second: Area):
-    __connections_left[first.area_id] += 1
-    __connections_left[second.area_id] += 1
+    origin = Transition(level_from, from_exit)
+    redirect = Transition(to_entrance, level_to)
+    create_connection(origin, redirect)
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:367]: fair enough. I didn't think of that.
Change made.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 27]
```diff
@@ -197,152 +222,280 @@ def highjack_transition_rando():
...
-        transitions_map[counterpart_original] = counterpart_redirect
-        origins_bucket.remove(counterpart_original)
-        redirections_bucket.remove(counterpart_redirect)
+    if can_reach_other_side(chosen_link, unchecked_links):
+        return can_reach_other_side(chosen_mirror, unchecked_links)
+    return False
+
+
+def find_and_break_open_connection(link_list: list[tuple[Transition, Transition]]):
+    direc = random.choice([-1, 1])
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:482]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:131]: To avoid having to iterate on all values twice, you could store the result of `",".join` in a variable and check if it's an empty string.
If all values are empty, `",".join([... if value is not None])` is equivalent to `",".join([])` which evaluates to `""`

The implementation can look something like this:
```py
def create_own_style(params: dict[str, str | None]):
    output_style = ",".join([
        f"&quot;{key}&quot;:&quot;{value}&quot;"
        for key, value in params.items()
        if value is not None
    ])
    return ' ownStyles="{&quot;0&quot;:{' + output_style + '}}"' if output_style else ""
```

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 28]
```diff
@@ -352,69 +505,102 @@ def get_random_redirection(original: Transition, all_redirections: Iterable[Tran
...
 
-        index = 2
+        global __current_hub, loose_ends, total_con_left
+
+        __current_hub = closed_door_levels[0]  # this list is shuffled, so just pick the first one
+        loose_ends = [*closed_door_exits]
+        total_con_left = sum(__connections_left[level] for level in closed_door_levels)
+
+        for index in range(1, len(closed_door_levels)):  # we skip the HUB, so we don't start at 0
+            direc = random.choice([-1, 1])
```
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:549]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 15 above)
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:545]: Didn't know that. Sure.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 16 above)
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:593]: The reason I don't let the program crash is because I want to be able to generate the graph anyway.
I use the graph to find out what is causing this crash.
What are your thoughts on this?

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:506]: The best way to keep PRs self-contained, would be to have a separate PR first that split it off.
Since you're already well started on this one, *if* we agree to have a small PR that does nothing else than split up this function right after it's merged, then you can add.
```suggestion
# TODO: Break up in smaller functions before changing anything else
def set_transitions_map():  # noqa: C901, PLR0912, PLR0914, PLR0915
```

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 17 above)
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:601]: I most certainly didn't know this one.
Personally I would prefer to keep it the old way because your proposed method doesn't feel intuitive to me at all, but hey you make the rules here I guess.
Changed.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:131]: Yeah that looks good, added!

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py [hunk 29]
```diff
@@ -197,152 +222,280 @@ def highjack_transition_rando():
...
-            ALL_POSSIBLE_TRANSITIONS_RANDO = [  # pyright: ignore[reportConstantRedefinition]
-                x for x in ALL_POSSIBLE_TRANSITIONS_RANDO if x != trans
-            ]
 
+def choose_random_exit(
+    level: int,
+    bucket_type: int,
+    priority: int,
+):
+    all_exits_available: list[int] = []
```
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:330]: Now assuming Python 3.11 https://github.com/Avasam/ptle-tools/pull/67 these checks can be made as match-case statements so that type-checkers ensure an exhaustive check. Or as simple if-else if you don't expect these enums to ever be more than 2 values

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:506]: Sure, I can agree to that.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 16 above)
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:595]: Ah that makes sense, that's useful. Let's add a comment explaining why we suppress the exception then, so I'm not tempted to move this when I forget why it's there :P

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:81]: Hmm... I'd rather not have the graph *depend* on `entrance_rando` (by separation of concerns, the main script *uses* the transition rando results as an input for the graph, and will do so for item rando, custom shaman shop, etc.). Before making any change, that brings up a design question:
Should the outposts be seen as different levels?
It's possible to traverse across with dash or pickaxe, meaning if they're seen as different levels, they should always be connected together on the graph. This question will also be relevant long term with item logic.
I also feel like the outposts only make sense being seen as different levels if the wells are also randomized (which we can't do rn either, because the "previous area" in-memory data isn't sufficient to spawn onto the right well (although there is a hacky workaround which is to simply *not* spawn on the well, and use the other 2 entrances instead)

[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py:81]: > To name one thing it allows to display `Burning Outpost` and `Jungle Outpost`, both of which aren't present in `TRANSITION_INFOS_DICT`, as that dict considers them both to be equal to `Twin Outposts`.

That's not implemented yet though, right?
image

[the remark this answers is not in the document]
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py]: - Yes the outposts should be seen as different levels, because if you don't you cause all kinds of problems + it's not intuitive at all for players that these are the same level, as it's not possible at all to go from the Flooded Courtyard exit to the Turtle Monument exit without using very specific glitches.
- If you as a player are crazy enough to traverse between the two using an item then I say good for you. This shouldn't impact the map generation whatsoever. The furthest I'm willing to go here is add an extra dotted line between the levels after all other generation is done, and even that's a stretch.
- I don't know what you mean by randomizing wells. What "wells"?
- About it not being implemented yet: I already have it implemented on my local version for weeks now. The only reason you don't have it yet is because you insist on smaller merge requests, and the merge requests we've done so far have been so slow.

So long story short: I thought about all of this and doing it this way was a very well thought out choice.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 16 above)
Comment (wossnameGitHub) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:584]: Done.

File: Dolphin scripts/Entrance Randomizer/lib/graph_creation.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Samuel-Therrien-Beslogic) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py]: By "wells" I mean the entrances to the underwater area (the entrance is inside of a well on both sides, that's what I was referring too).

I guess they could always be marked separately on the graph with "unrandomized" connection in between, or w/e visual marker, now that that's a thing. So less of a concern.

> About it not being implemented yet: I already have it implemented on my local version

Ok I thought I missed something in this PR.

To answer your question on Discord about fixing the newly appeared pyright issue due to the name containing an underscore:
It simply shows in a more obvious way the "leak" of responsibility between `entrance_rando` and `graph_creation`.
1. You *could* do the necessary code modification to solve that with in mind whatever other changed you have prepped that I'm not necessarily aware of. But neither you nor I would like to add that to this PR.
2. You could go back to using `TRANSITION_INFOS_DICT`, which currently has the exact same effect, reduces changes here (ie: the things you mentioned about Twin Outpost is not of concern for this PR, but is good to know for later). And the need for this data can be addressed with the PR that adds it.
3. You can add `# pyright: ignore [theErrorCode]` to get this PR green, and I'll just revert it to `TRANSITION_INFOS_DICT` after anyway, so like 2. the needs for this data can be addressed in the proper PR.

[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/graph_creation.py]: By "wells" I mean the entrances to the underwater area (the entrance is inside of a well on both sides, that's what I was referring too).

I guess they could always be marked separately on the graph with "unrandomized" connection in between, or w/e visual marker, now that that's a thing. So less of a concern.

> About it not being implemented yet: I already have it implemented on my local version

Ok I thought I missed something in this PR.

To answer your question on Discord about fixing the newly appeared pyright issue due to the name containing an underscore:
It simply shows in a more obvious way the "leak" of responsibility between `entrance_rando` and `graph_creation`.
1. You *could* do the necessary code modification to solve that with in mind whatever other changed you have prepped that I'm not necessarily aware of. But neither you nor I would like to add that to this PR.
2. You could go back to using `TRANSITION_INFOS_DICT`, which currently has the exact same effect, reduces changes here (ie: the things you mentioned about Twin Outpost is not of concern for this PR, but is good to know for later). And the need for this data can be addressed with the PR that adds it.
3. You can add `# pyright: ignore [theErrorCode]` to get this PR green, and I'll just revert it to `TRANSITION_INFOS_DICT` after anyway, so like 2. the needs for this data can be addressed in the proper PR.

File: Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py:316]: Just to be extra clear: I'm not asking you to make a change here.

**Review by Avasam — APPROVED** [review]
